File: packages/react/src/animation.ts
// Frame-driven animation helpers, meant to be combined with useCurrentFrame()
// (hooks.ts). Neither function here reads React state itself — both are
// plain math over whatever frame/time value the caller passes in, so they
// work the same whether called from a component body or anywhere else.

export type Extrapolate = 'extend' | 'clamp' | 'identity';

export interface InterpolateOptions {
  easing?: (input: number) => number;
  extrapolateLeft?: Extrapolate;
  extrapolateRight?: Extrapolate;
}

/**
 * Maps `input` from `inputRange` into `outputRange`, linearly interpolating
 * within each segment (and applying `options.easing` to the 0-1 position
 * inside that segment, if given). `inputRange` must be strictly increasing
 * and the same length as `outputRange` (at least 2 entries each).
 *
 * Outside `inputRange`, `extrapolateLeft`/`extrapolateRight` control the
 * result: `'extend'` (default) continues the boundary segment's linear
 * slope, `'clamp'` holds the boundary output value, and `'identity'`
 * returns `input` itself unchanged.
 */
export function interpolate(
  input: number,
  inputRange: readonly number[],
  outputRange: readonly number[],
  options: InterpolateOptions = {},
): number {
  if (inputRange.length < 2 || inputRange.length !== outputRange.length) {
    throw new Error('interpolate() requires inputRange and outputRange of the same length, at least 2');
  }
  for (let i = 1; i < inputRange.length; i += 1) {
    if (inputRange[i] <= inputRange[i - 1]) {
      throw new Error('interpolate() requires inputRange to be strictly increasing');
    }
  }

  const { easing = (t: number) => t, extrapolateLeft = 'extend', extrapolateRight = 'extend' } = options;

  if (input < inputRange[0]) {
    if (extrapolateLeft === 'identity') {
      return input;
    }
    if (extrapolateLeft === 'clamp') {
      return outputRange[0];
    }
  }
  const lastIndex = inputRange.length - 1;
  if (input > inputRange[lastIndex]) {
    if (extrapolateRight === 'identity') {
      return input;
    }
    if (extrapolateRight === 'clamp') {
      return outputRange[lastIndex];
    }
  }

  let segment = 0;
  while (segment < inputRange.length - 2 && input >= inputRange[segment + 1]) {
    segment += 1;
  }
  const segmentStart = inputRange[segment];
  const segmentEnd = inputRange[segment + 1];
  const progress = easing((input - segmentStart) / (segmentEnd - segmentStart));
  return outputRange[segment] + progress * (outputRange[segment + 1] - outputRange[segment]);
}

/**
 * Common easing curves for `interpolate()`'s `easing` option. Named `Easings`
 * (not `Easing`) to avoid colliding with the generated `Easing` union type
 * used by the project's own `Animatable<T>`/`Keyframe<T>` model, which is an
 * unrelated, project.json-facing concept.
 */
export const Easings = {
  linear: (t: number): number => t,
  easeIn: (t: number): number => t * t,
  easeOut: (t: number): number => t * (2 - t),
  easeInOut: (t: number): number => (t < 0.5 ? 2 * t * t : -1 + (4 - 2 * t) * t),
  easeInSine: (t: number): number => 1 - Math.cos((t * Math.PI) / 2),
  easeOutSine: (t: number): number => Math.sin((t * Math.PI) / 2),
  easeInOutSine: (t: number): number => -(Math.cos(Math.PI * t) - 1) / 2,
  easeInQuad: (t: number): number => t * t,
  easeOutQuad: (t: number): number => t * (2 - t),
  easeInOutQuad: (t: number): number => (t < 0.5 ? 2 * t * t : -1 + (4 - 2 * t) * t),
  easeInCubic: (t: number): number => t * t * t,
  easeOutCubic: (t: number): number => --t * t * t + 1,
  easeInOutCubic: (t: number): number => (t < 0.5 ? 4 * t * t * t : (t - 1) * (2 * t - 2) * (2 * t - 2) + 1),
  easeInQuart: (t: number): number => t * t * t * t,
  easeOutQuart: (t: number): number => 1 - --t * t * t * t,
  easeInOutQuart: (t: number): number => (t < 0.5 ? 8 * t * t * t * t : 1 - 8 * --t * t * t * t),
  easeInQuint: (t: number): number => t * t * t * t * t,
  easeOutQuint: (t: number): number => 1 + --t * t * t * t * t,
  easeInOutQuint: (t: number): number => (t < 0.5 ? 16 * t * t * t * t * t : 1 + 16 * --t * t * t * t * t),
  easeInExpo: (t: number): number => (t === 0 ? 0 : Math.pow(1024, t - 1)),
  easeOutExpo: (t: number): number => (t === 1 ? 1 : 1 - Math.pow(2, -10 * t)),
  easeInOutExpo: (t: number): number =>
    t === 0 ? 0 : t === 1 ? 1 : t < 0.5 ? Math.pow(1024, t * 2 - 1) / 2 : (2 - Math.pow(2, -20 * t + 10)) / 2,
  easeInCirc: (t: number): number => 1 - Math.sqrt(1 - t * t),
  easeOutCirc: (t: number): number => Math.sqrt(1 - --t * t),
  easeInOutCirc: (t: number): number =>
    t < 0.5 ? (1 - Math.sqrt(1 - 4 * t * t)) / 2 : (Math.sqrt(1 - --t * 2 * t * 2) + 1) / 2,
  easeInBack: (t: number): number => {
    const c1 = 1.70158;
    const c3 = c1 + 1;
    return c3 * t * t * t - c1 * t * t;
  },
  easeOutBack: (t: number): number => {
    const c1 = 1.70158;
    const c3 = c1 + 1;
    return 1 + c3 * --t * t * t + c1 * t * t;
  },
  easeInOutBack: (t: number): number => {
    const c1 = 1.70158;
    const c2 = c1 * 1.525;
    return t < 0.5
      ? (Math.pow(2 * t, 2) * ((c2 + 1) * 2 * t - c2)) / 2
      : (Math.pow(2 * t - 2, 2) * ((c2 + 1) * (t * 2 - 2) + c2) + 2) / 2;
  },
  easeInElastic: (t: number): number => {
    const c4 = (2 * Math.PI) / 3;
    return t === 0
      ? 0
      : t === 1
        ? 1
        : -Math.pow(2, 10 * t - 10) * Math.sin((t * 10 - 10.75) * c4);
  },
  easeOutElastic: (t: number): number => {
    const c4 = (2 * Math.PI) / 3;
    return t === 0
      ? 0
      : t === 1
        ? 1
        : Math.pow(2, -10 * t) * Math.sin((t * 10 - 0.75) * c4) + 1;
  },
  easeInOutElastic: (t: number): number => {
    const c5 = (2 * Math.PI) / 4.5;
    return t === 0
      ? 0
      : t === 1
        ? 1
        : t < 0.5
          ? -(Math.pow(2, 20 * t - 10) * Math.sin((20 * t - 11.125) * c5)) / 2
          : (Math.pow(2, -20 * t + 10) * Math.sin((20 * t - 11.125) * c5)) / 2 + 1;
  },
  easeInBounce: (t: number): number => 1 - Easings.easeOutBounce(1 - t),
  easeOutBounce: (t: number): number => {
    const n1 = 7.5625;
    const d1 = 2.75;
    if (t < 1 / d1) {
      return n1 * t * t;
    } else if (t < 2 / d1) {
      return n1 * (t -= 1.5 / d1) * t + 0.75;
    } else if (t < 2.5 / d1) {
      return n1 * (t -= 2.25 / d1) * t + 0.9375;
    } else {
      return n1 * (t -= 2.625 / d1) * t + 0.984375;
    }
  },
  easeInOutBounce: (t: number): number =>
    t < 0.5 ? (1 - Easings.easeOutBounce(1 - 2 * t)) / 2 : (1 + Easings.easeOutBounce(2 * t - 1)) / 2,
};

export interface SpringConfig {
  damping?: number;
  mass?: number;
  stiffness?: number;
  /** Hold the settled value once it's reached instead of overshooting past it. */
  overshootClamping?: boolean;
}

export interface SpringOptions {
  frame: number;
  fps: number;
  config?: SpringConfig;
  from?: number;
  to?: number;
  delay?: number;
  /** Frames beyond this are clamped to the settled value. */
  durationInFrames?: number;
}

const DEFAULT_SPRING_CONFIG: Required<SpringConfig> = {
  damping: 10,
  mass: 1,
  stiffness: 100,
  overshootClamping: false,
};

/**
 * A damped harmonic oscillator's step response (0 at t=0 settling toward 1),
 * mapped into `[from, to]`. `frame`/`fps`/`delay` convert to elapsed seconds;
 * frames before `delay` return `from`.
 */
export function spring(options: SpringOptions): number {
  const { frame, fps, from = 0, to = 1, delay = 0 } = options;
  if (fps <= 0) {
    throw new Error('spring() requires a positive fps');
  }
  const config = { ...DEFAULT_SPRING_CONFIG, ...options.config };

  let effectiveFrame = frame - delay;
  if (options.durationInFrames !== undefined) {
    effectiveFrame = Math.min(effectiveFrame, options.durationInFrames - delay);
  }
  if (effectiveFrame <= 0) {
    return from;
  }

  const t = effectiveFrame / fps;
  const settled = dampedOscillatorStepResponse(t, config);
  const value = from + settled * (to - from);

  if (config.overshootClamping) {
    return to >= from ? Math.min(value, to) : Math.max(value, to);
  }
  return value;
}

/** Step response x(t) of m*x'' + c*x' + k*x = k, x(0) = 0, x'(0) = 0. */
function dampedOscillatorStepResponse(t: number, config: Required<SpringConfig>): number {
  const { mass, stiffness, damping } = config;
  const naturalFrequency = Math.sqrt(stiffness / mass);
  const dampingRatio = damping / (2 * Math.sqrt(stiffness * mass));

  if (dampingRatio < 1) {
    const dampedFrequency = naturalFrequency * Math.sqrt(1 - dampingRatio * dampingRatio);
    const envelope = Math.exp(-dampingRatio * naturalFrequency * t);
    return (
      1 -
      envelope *
      (Math.cos(dampedFrequency * t) +
        ((dampingRatio * naturalFrequency) / dampedFrequency) * Math.sin(dampedFrequency * t))
    );
  }
  if (dampingRatio === 1) {
    return 1 - Math.exp(-naturalFrequency * t) * (1 + naturalFrequency * t);
  }
  const root = naturalFrequency * Math.sqrt(dampingRatio * dampingRatio - 1);
  const r1 = -naturalFrequency * dampingRatio + root;
  const r2 = -naturalFrequency * dampingRatio - root;
  return 1 - (r2 * Math.exp(r1 * t) - r1 * Math.exp(r2 * t)) / (r2 - r1);
}

File: crates/composition/src/animation.rs
use std::error::Error;
use std::fmt;

use crate::{Animatable, Easing, Time, TimeError};

/// Evaluates a numeric property at item-local `time`.
///
/// Easing belongs to the destination keyframe for the segment ending there.
/// Values before or after the animation are clamped to its first or last value.
pub fn evaluate_f64(value: &Animatable<f64>, time: Time) -> Result<f64, AnimationError> {
    let Animatable::Keyframes(animation) = value else {
        let Animatable::Static(value) = value else {
            unreachable!()
        };
        return Ok(*value);
    };

    let first = animation
        .keyframes
        .first()
        .ok_or(AnimationError::EmptyKeyframes)?;
    if time.cmp_exact(first.time)?.is_le() {
        return Ok(first.value);
    }

    for pair in animation.keyframes.windows(2) {
        let [from, to] = pair else { unreachable!() };
        if time.cmp_exact(to.time)?.is_le() {
            let elapsed = time.checked_sub(from.time)?.as_seconds()?;
            let duration = to.time.checked_sub(from.time)?.as_seconds()?;
            if duration <= 0.0 {
                return Ok(to.value);
            }
            let progress = (elapsed / duration).clamp(0.0, 1.0);
            let eased = apply_easing(progress, to.easing.unwrap_or(Easing::Linear));
            return Ok(from.value + (to.value - from.value) * eased);
        }
    }

    Ok(animation
        .keyframes
        .last()
        .expect("first keyframe exists")
        .value)
}

/// Integrates a numeric animation from item-local zero through `time`.
///
/// This is used to turn animated playback rate into an absolute source time.
pub fn integrate_f64(value: &Animatable<f64>, time: Time) -> Result<f64, AnimationError> {
    let end = time.as_seconds()?.max(0.0);
    let Animatable::Keyframes(animation) = value else {
        let Animatable::Static(value) = value else {
            unreachable!()
        };
        return Ok(*value * end);
    };
    let first = animation
        .keyframes
        .first()
        .ok_or(AnimationError::EmptyKeyframes)?;
    let first_time = first.time.as_seconds()?.max(0.0);
    let mut integral = first.value * end.min(first_time);
    if end <= first_time {
        return Ok(integral);
    }

    for pair in animation.keyframes.windows(2) {
        let [from, to] = pair else { unreachable!() };
        let start = from.time.as_seconds()?.max(0.0);
        let finish = to.time.as_seconds()?.max(start);
        let duration = finish - start;
        if duration == 0.0 {
            continue;
        }
        let progress = ((end - start) / duration).clamp(0.0, 1.0);
        let easing = to.easing.unwrap_or(Easing::Linear);
        integral += duration
            * (from.value * progress + (to.value - from.value) * easing_integral(progress, easing));
        if end <= finish {
            return Ok(integral);
        }
    }

    let last = animation.keyframes.last().expect("first keyframe exists");
    let last_time = last.time.as_seconds()?.max(0.0);
    integral += last.value * (end - last_time).max(0.0);
    Ok(integral)
}

/// The easings.net catalogue, mirroring
/// `packages/react/src/animation.ts`'s `Easings` object formula-for-formula
/// so the same named curve looks identical whether it drives a project.json
/// keyframe (here) or an `interpolate()` call in a React composition.
fn apply_easing(progress: f64, easing: Easing) -> f64 {
    let t = progress;
    match easing {
        Easing::Linear => t,
        Easing::EaseIn => t * t,
        Easing::EaseOut => 1.0 - (1.0 - t) * (1.0 - t),
        Easing::EaseInOut => t * t * (3.0 - 2.0 * t),
        Easing::EaseInSine => 1.0 - (t * std::f64::consts::FRAC_PI_2).cos(),
        Easing::EaseOutSine => (t * std::f64::consts::FRAC_PI_2).sin(),
        Easing::EaseInOutSine => -((std::f64::consts::PI * t).cos() - 1.0) / 2.0,
        Easing::EaseInQuad => t * t,
        Easing::EaseOutQuad => t * (2.0 - t),
        Easing::EaseInOutQuad => {
            if t < 0.5 {
                2.0 * t * t
            } else {
                -1.0 + (4.0 - 2.0 * t) * t
            }
        }
        Easing::EaseInCubic => t * t * t,
        Easing::EaseOutCubic => (t - 1.0).powi(3) + 1.0,
        Easing::EaseInOutCubic => {
            if t < 0.5 {
                4.0 * t * t * t
            } else {
                (t - 1.0) * (2.0 * t - 2.0).powi(2) + 1.0
            }
        }
        Easing::EaseInQuart => t.powi(4),
        Easing::EaseOutQuart => 1.0 - (t - 1.0).powi(4),
        Easing::EaseInOutQuart => {
            if t < 0.5 {
                8.0 * t.powi(4)
            } else {
                1.0 - 8.0 * (t - 1.0).powi(4)
            }
        }
        Easing::EaseInQuint => t.powi(5),
        Easing::EaseOutQuint => 1.0 + (t - 1.0).powi(5),
        Easing::EaseInOutQuint => {
            if t < 0.5 {
                16.0 * t.powi(5)
            } else {
                1.0 + 16.0 * (t - 1.0).powi(5)
            }
        }
        Easing::EaseInExpo => {
            if t == 0.0 {
                0.0
            } else {
                2.0_f64.powf(10.0 * t - 10.0)
            }
        }
        Easing::EaseOutExpo => {
            if t == 1.0 {
                1.0
            } else {
                1.0 - 2.0_f64.powf(-10.0 * t)
            }
        }
        Easing::EaseInOutExpo => {
            if t == 0.0 {
                0.0
            } else if t == 1.0 {
                1.0
            } else if t < 0.5 {
                2.0_f64.powf(20.0 * t - 10.0) / 2.0
            } else {
                (2.0 - 2.0_f64.powf(-20.0 * t + 10.0)) / 2.0
            }
        }
        Easing::EaseInCirc => 1.0 - (1.0 - t * t).sqrt(),
        Easing::EaseOutCirc => (1.0 - (t - 1.0).powi(2)).sqrt(),
        Easing::EaseInOutCirc => {
            if t < 0.5 {
                (1.0 - (1.0 - 4.0 * t * t).sqrt()) / 2.0
            } else {
                ((1.0 - 4.0 * (t - 1.0).powi(2)).sqrt() + 1.0) / 2.0
            }
        }
        Easing::EaseInBack => {
            const C1: f64 = 1.70158;
            const C3: f64 = C1 + 1.0;
            C3 * t * t * t - C1 * t * t
        }
        Easing::EaseOutBack => {
            const C1: f64 = 1.70158;
            const C3: f64 = C1 + 1.0;
            1.0 + C3 * (t - 1.0).powi(3) + C1 * (t - 1.0).powi(2)
        }
        Easing::EaseInOutBack => {
            const C1: f64 = 1.70158;
            const C2: f64 = C1 * 1.525;
            if t < 0.5 {
                (2.0 * t).powi(2) * ((C2 + 1.0) * 2.0 * t - C2) / 2.0
            } else {
                ((2.0 * t - 2.0).powi(2) * ((C2 + 1.0) * (t * 2.0 - 2.0) + C2) + 2.0) / 2.0
            }
        }
        Easing::EaseInElastic => {
            const C4: f64 = 2.0 * std::f64::consts::PI / 3.0;
            if t == 0.0 {
                0.0
            } else if t == 1.0 {
                1.0
            } else {
                -(2.0_f64.powf(10.0 * t - 10.0)) * ((t * 10.0 - 10.75) * C4).sin()
            }
        }
        Easing::EaseOutElastic => {
            const C4: f64 = 2.0 * std::f64::consts::PI / 3.0;
            if t == 0.0 {
                0.0
            } else if t == 1.0 {
                1.0
            } else {
                2.0_f64.powf(-10.0 * t) * ((t * 10.0 - 0.75) * C4).sin() + 1.0
            }
        }
        Easing::EaseInOutElastic => {
            const C5: f64 = 2.0 * std::f64::consts::PI / 4.5;
            if t == 0.0 {
                0.0
            } else if t == 1.0 {
                1.0
            } else if t < 0.5 {
                -(2.0_f64.powf(20.0 * t - 10.0) * ((20.0 * t - 11.125) * C5).sin()) / 2.0
            } else {
                (2.0_f64.powf(-20.0 * t + 10.0) * ((20.0 * t - 11.125) * C5).sin()) / 2.0 + 1.0
            }
        }
        Easing::EaseInBounce => 1.0 - ease_out_bounce(1.0 - t),
        Easing::EaseOutBounce => ease_out_bounce(t),
        Easing::EaseInOutBounce => {
            if t < 0.5 {
                (1.0 - ease_out_bounce(1.0 - 2.0 * t)) / 2.0
            } else {
                (1.0 + ease_out_bounce(2.0 * t - 1.0)) / 2.0
            }
        }
    }
}

/// Shared by `EaseOutBounce`/`EaseInBounce`/`EaseInOutBounce`, mirroring
/// `Easings.easeOutBounce` in `packages/react/src/animation.ts`.
fn ease_out_bounce(t: f64) -> f64 {
    const N1: f64 = 7.5625;
    const D1: f64 = 2.75;
    if t < 1.0 / D1 {
        N1 * t * t
    } else if t < 2.0 / D1 {
        let t = t - 1.5 / D1;
        N1 * t * t + 0.75
    } else if t < 2.5 / D1 {
        let t = t - 2.25 / D1;
        N1 * t * t + 0.9375
    } else {
        let t = t - 2.625 / D1;
        N1 * t * t + 0.984375
    }
}

fn easing_integral(progress: f64, easing: Easing) -> f64 {
    match easing {
        Easing::Linear => progress * progress / 2.0,
        Easing::EaseIn => progress * progress * progress / 3.0,
        Easing::EaseOut => progress * progress - progress * progress * progress / 3.0,
        Easing::EaseInOut => progress * progress * progress - progress.powi(4) / 2.0,
        // The rest of the catalogue (trig, exponential, and multi-branch
        // piecewise curves like elastic/back/bounce) doesn't have an
        // antiderivative worth hand-deriving and hand-verifying by eye the
        // way the four polynomial curves above do — `integrate_f64`'s own
        // caller only needs this to convert an animated playback rate into
        // an absolute source-time offset, so a numerically integrated
        // approximation (indistinguishable from the closed form at
        // f64 precision for these smooth curves) is the pragmatic choice.
        _ => integrate_numerically(progress, |p| apply_easing(p, easing)),
    }
}

/// Composite Simpson's rule for ∫₀^progress f(x) dx, `progress` assumed
/// non-negative (always true for the `[0, 1]`-clamped progress values this
/// module calls it with).
fn integrate_numerically(progress: f64, f: impl Fn(f64) -> f64) -> f64 {
    if progress <= 0.0 {
        return 0.0;
    }
    const STEPS: usize = 128; // even, required by Simpson's rule
    let h = progress / STEPS as f64;
    let mut sum = f(0.0) + f(progress);
    for i in 1..STEPS {
        let x = i as f64 * h;
        sum += f(x) * if i % 2 == 0 { 2.0 } else { 4.0 };
    }
    sum * h / 3.0
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum AnimationError {
    EmptyKeyframes,
    Time(TimeError),
}

impl fmt::Display for AnimationError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::EmptyKeyframes => formatter.write_str("animation has no keyframes"),
            Self::Time(error) => write!(formatter, "could not evaluate animation time: {error}"),
        }
    }
}

impl Error for AnimationError {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        match self {
            Self::EmptyKeyframes => None,
            Self::Time(error) => Some(error),
        }
    }
}

impl From<TimeError> for AnimationError {
    fn from(error: TimeError) -> Self {
        Self::Time(error)
    }
}

#[cfg(test)]
mod tests {
    use crate::{Easing, Keyframe, KeyframeAnimation, KeyframeAnimationType};

    use super::*;

    #[test]
    fn every_easing_starts_at_0_and_ends_at_1() {
        // t=0/t=1 are exact for every curve in the catalogue (including the
        // ones that special-case those endpoints, like elastic/expo, to
        // avoid landing exactly on an asymptote).
        for easing in ALL_EASINGS {
            assert!((apply_easing(0.0, easing) - 0.0).abs() < 1e-9, "{easing:?} at t=0");
            assert!((apply_easing(1.0, easing) - 1.0).abs() < 1e-9, "{easing:?} at t=1");
        }
    }

    #[test]
    fn ease_in_out_variants_pass_through_the_midpoint() {
        // Every "InOut" curve is point-symmetric about (0.5, 0.5).
        for easing in [
            Easing::EaseInOut,
            Easing::EaseInOutSine,
            Easing::EaseInOutQuad,
            Easing::EaseInOutCubic,
            Easing::EaseInOutQuart,
            Easing::EaseInOutQuint,
            Easing::EaseInOutCirc,
            Easing::EaseInOutBack,
            Easing::EaseInOutBounce,
        ] {
            assert!(
                (apply_easing(0.5, easing) - 0.5).abs() < 1e-9,
                "{easing:?} at t=0.5"
            );
        }
    }

    #[test]
    fn easing_integral_matches_the_closed_form_for_the_legacy_curves() {
        // The numeric fallback (used for everything but the four original
        // curves) should agree with the hand-derived closed form for those
        // four, cross-checking `integrate_numerically` itself.
        for easing in [Easing::Linear, Easing::EaseIn, Easing::EaseOut, Easing::EaseInOut] {
            let closed_form = easing_integral(0.7, easing);
            let numeric = integrate_numerically(0.7, |p| apply_easing(p, easing));
            assert!(
                (closed_form - numeric).abs() < 1e-6,
                "{easing:?}: closed form {closed_form} vs numeric {numeric}"
            );
        }
    }

    #[test]
    fn easing_integral_is_monotonic_and_bounded_for_every_curve() {
        // A loose sanity check on the numeric fallback for curves that
        // overshoot/undershoot [0, 1] (back, elastic): the integral up to
        // t=1 should still land close to the curve's actual signed area,
        // which for all of these is within a small multiple of the [0, 1]
        // range front-loaded/back-loaded eases live in.
        for easing in ALL_EASINGS {
            let total = easing_integral(1.0, easing);
            assert!(total.is_finite(), "{easing:?} integral is finite");
        }
    }

    const ALL_EASINGS: [Easing; 34] = [
        Easing::Linear,
        Easing::EaseIn,
        Easing::EaseOut,
        Easing::EaseInOut,
        Easing::EaseInSine,
        Easing::EaseOutSine,
        Easing::EaseInOutSine,
        Easing::EaseInQuad,
        Easing::EaseOutQuad,
        Easing::EaseInOutQuad,
        Easing::EaseInCubic,
        Easing::EaseOutCubic,
        Easing::EaseInOutCubic,
        Easing::EaseInQuart,
        Easing::EaseOutQuart,
        Easing::EaseInOutQuart,
        Easing::EaseInQuint,
        Easing::EaseOutQuint,
        Easing::EaseInOutQuint,
        Easing::EaseInExpo,
        Easing::EaseOutExpo,
        Easing::EaseInOutExpo,
        Easing::EaseInCirc,
        Easing::EaseOutCirc,
        Easing::EaseInOutCirc,
        Easing::EaseInBack,
        Easing::EaseOutBack,
        Easing::EaseInOutBack,
        Easing::EaseInElastic,
        Easing::EaseOutElastic,
        Easing::EaseInOutElastic,
        Easing::EaseInBounce,
        Easing::EaseOutBounce,
        Easing::EaseInOutBounce,
    ];

    #[test]
    fn interpolates_using_item_local_time() {
        let animation = Animatable::Keyframes(KeyframeAnimation {
            kind: KeyframeAnimationType::Keyframes,
            keyframes: vec![
                Keyframe {
                    time: Time::ZERO,
                    value: 0.0,
                    easing: None,
                },
                Keyframe {
                    time: Time::new(1, 1),
                    value: 100.0,
                    easing: Some(Easing::Linear),
                },
            ],
        });

        assert_eq!(evaluate_f64(&animation, Time::new(1, 2)).unwrap(), 50.0);
    }

    #[test]
    fn clamps_outside_the_keyframe_range() {
        let animation = Animatable::Keyframes(KeyframeAnimation {
            kind: KeyframeAnimationType::Keyframes,
            keyframes: vec![Keyframe {
                time: Time::new(1, 1),
                value: 42.0,
                easing: None,
            }],
        });

        assert_eq!(evaluate_f64(&animation, Time::ZERO).unwrap(), 42.0);
        assert_eq!(evaluate_f64(&animation, Time::new(2, 1)).unwrap(), 42.0);
    }

    #[test]
    fn integrates_playback_rate_across_keyframes() {
        let animation = Animatable::Keyframes(KeyframeAnimation {
            kind: KeyframeAnimationType::Keyframes,
            keyframes: vec![
                Keyframe {
                    time: Time::ZERO,
                    value: 1.0,
                    easing: None,
                },
                Keyframe {
                    time: Time::new(2, 1),
                    value: 3.0,
                    easing: Some(Easing::Linear),
                },
            ],
        });

        assert_eq!(integrate_f64(&animation, Time::new(2, 1)).unwrap(), 4.0);
        assert_eq!(integrate_f64(&animation, Time::new(3, 1)).unwrap(), 7.0);
    }
}
